UefiTestingPkg/AuditTests/PagingAudit/UEFI/Arm64/PagingAuditProcessor.c

```diff
@@ -0,0 +1,240 @@
+            Pte4K = (UINT64 *)(Pte2M[Index2] & TT_ADDRESS_MASK);
+            MyPdeCount++;
+
+            if (MyPdeCount <= *PdeCount) {
+              PdeEntries[MyPdeCount-1] = (UINT64)Pte4K;
+            }
+
+            for (Index3 = 0x0; Index3 < TT_ENTRY_COUNT; Index3++ ) {
+              if (!IS_BLOCK (Pte4K[Index3], 3)) {
+                NumPage4KNotPresent++;
+                Address = IndexToAddress (Index0, Index1, Index2, Index3);
+                if ((mMemoryProtectionProtocol != NULL) && (mMemoryProtectionProtocol->IsGuardPage (Address))) {
+                  MyGuardCount++;
+                  if (MyGuardCount <= *GuardCount) {
+                    GuardEntries[MyGuardCount - 1] = Address;
+                  }
+
+                  continue;
+                }
+              }
+
+              My4KCount++;
+              if (My4KCount <= *Pte4KCount) {
+                Pte4KEntries[My4KCount-1] = Pte4K[Index3] | IndexToAddress (Index0, Index1, Index2, Index3);
+              }
+            }
+          } else {
+            My2MCount++;
+
+            if (My2MCount <= *Pte2MCount) {
+              Pte2MEntries[My2MCount-1] = Pte2M[Index2] | IndexToAddress (Index0, Index1, Index2, Index3);
+            }
+          }
+        }
+      } else {
+        My1GCount++;
+
+        if (My1GCount <= *Pte1GCount) {
+          Pte1GEntries[My1GCount-1] = Pte1G[Index1] | IndexToAddress (Index0, Index1, Index2, Index3);
+        }
+      }
+    }
+  }
+
+  DEBUG ((DEBUG_ERROR, "Pages used for Page Tables   = %d\n", MyPdeCount));
+  DEBUG ((DEBUG_ERROR, "Number of   4K Pages active  = %d - NotPresent = %d\n", My4KCount - NumPage4KNotPresent, NumPage4KNotPresent));
+  DEBUG ((DEBUG_ERROR, "Number of   2M Pages active  = %d - NotPresent = %d\n", My2MCount - NumPage2MNotPresent, NumPage2MNotPresent));
+  DEBUG ((DEBUG_ERROR, "Number of   1G Pages active  = %d - NotPresent = %d\n", My1GCount - NumPage1GNotPresent, NumPage1GNotPresent));
+  DEBUG ((DEBUG_ERROR, "Number of   Guard Pages active  = %d\n", MyGuardCount));
+
+  //
+  // determine whether any of the buffers were too small.
+  // Only matters if a given buffer was provided.
+  //
+  if (((Pte1GEntries != NULL) && (*Pte1GCount < My1GCount)) || ((Pte2MEntries != NULL) && (*Pte2MCount < My2MCount)) ||
+      ((Pte4KEntries != NULL) && (*Pte4KCount < My4KCount)) || ((PdeEntries != NULL) && (*PdeCount < MyPdeCount)) ||
+      ((GuardEntries != NULL) && (*GuardCount < MyGuardCount)))
+  {
+    Status = EFI_BUFFER_TOO_SMALL;
+  }
+
+  //
+  // Update all the return pointers.
+  //
+  *Pte1GCount = My1GCount;
+  *Pte2MCount = My2MCount;
+  *Pte4KCount = My4KCount;
+  *PdeCount   = MyPdeCount;
+  *GuardCount = MyGuardCount;
+
+  return Status;
+}
+
+/**
+  Calculate the maximum physical address bits supported.
+
+  @return the maximum support physical address bits supported.
+**/
+UINT8
+EFIAPI
+CalculateMaximumSupportAddressBits (
+  VOID
+  )
+{
+  return 36;

```

Comment:
Why is this 36?

Reply:
You may be confusing total width with address width. The width of the address is 36 bits because the first 12 bits are unnecessary, so the address mask is 0xFFFFFFFFF000 giving 256 terabytes of addressable memory.

Reply:
I should clarify I'm just referring to the translation table when saying that. Writing to address 0x1000 is different from writing to address 0x1001, but both use the same translation table entry.

Reply:
Gotcha, thanks!